### PR TITLE
[Merged by Bors] - chore(Tactic): command/tactic/term elaborator for adaptation notes

### PR DESCRIPTION
**Pull request by jcommelin**
This PR introduces `#adaptation_note`. The main benefit is that this prevents typos in the string "Adaptation note". Such typos would make it hard to find the note later. A side benefit is that the command comes with a docstring that explains the purpose of adaptation notes in general.

Co-authored-by: Kyle Miller <kmill31415@gmail.com>

---

A follow-up PR can migrate all existing adaptation notes to the new syntax.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/AdaptationNote.lean [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    let doc := doc.updateTrailing trailing
+    let stx' := (← getRef)
+    let stx' := stx'.setArg 0 stx'[0].unsetTrailing
+    let stx' := stx'.setArg 1 (mkNullNode #[doc])
+    Meta.Tactic.TryThis.addSuggestion (← getRef) (f stx') (origSpan? := ← getRef)
+
+/-- Adaptation notes are comments that are used to indicate that a piece of code
+has been changed to accomodate a change in Lean core.
+They typically require further action/maintenance to be taken in the future. -/
+elab "#adaptation_note " (docComment)? : command => do
```
Comment (adomani) [inline Mathlib/Tactic/AdaptationNote.lean:42]:
```suggestion
elab (name := adaptationNoteCmd) "#adaptation_note " (docComment)? : command => do
```

If you name this here...

File: Mathlib/Tactic/AdaptationNote.lean [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  Elab.Command.liftTermElabM <| reportAdaptationNote (fun s => (⟨s⟩ : TSyntax `tactic))
+
+/-- Adaptation notes are comments that are used to indicate that a piece of code
+has been changed to accomodate a change in Lean core.
+They typically require further action/maintenance to be taken in the future. -/
```
Comment (adomani) [inline Mathlib/Tactic/AdaptationNote.lean:47]:
```suggestion
@[inherit_doc adaptationNoteCmd]
```
... you should be able to do this!

I might have gotten the syntax slightly wrong, this is untested!

**Review by adomani — APPROVED** [review]

**Comment by semorrison** [comment]
Damiano's suggestion looks good, otherwise

bors d+

**Comment by grunweg** [comment]
The "find #-command" linter fails, since it does not recognise #adaptation_note.
@adomani As the author of that lint, do you have a moment to help here?

**Comment by adomani** [comment]
#12675 should make the linter allow adaptation notes, untested.  Feel free to merge into this one!

**Comment by adomani** [comment]
Should adaptation notes also have a `since <date>` optional syntax, like `deprecate`?

**Comment by jcommelin** [comment]
@adomani Thanks for the linter fix. I merged it into this PR. Also thanks for the other suggestions. I'll look into them asap.

Concerning the optional `since`. I think that is a good idea. I suggest we do it as a follow-up on this PR.

**Comment by jcommelin** [comment]
bors merge

**Pull request closed without merging**